include/deal.II/numerics/vector_tools.h

```diff
@@ -1381,6 +1381,112 @@ namespace VectorTools
+   * 
+   * If the ConstraintMatrix @p constraints contained values or other
+   * constraints before, the new ones are added or the old ones overwritten,
+   * if a node of the boundary part to be used was already in the list of
+   * constraints. This is handled by using inhomogeneous constraints. Please
+   * note that when combining adaptive meshes and this kind of constraints, the
+   * Dirichlet conditions should be set first, and then completed by hanging
+   * node constraints, in order to make sure that the discretization remains
+   * consistent. See the discussion on conflicting constraints in the
+   * module on @ref constraints
+   * 
+   * <h4>Arguments to this function></h4>
+   *
+   * This function is explicitly for use with FE_Nedelec elements, or with FESystem 
+   * elements which contain FE_Nedelec elements. It will throw an exception if called
+   * with any other finite element. The user must ensure that FESystem elements are
+   * correctly setup when using this function as this check not possible in this case.
+   * 
+   * The second argument of this function denotes the first vector component of the 
+   * finite element which corresponds to the vector function that you wish to constrain.
+   * For example, if we are solving Maxwell's equations in 3D and have components 
+   * $(E_x,E_y,E_z,B_x,B_y,B_z)$ and we want the boundary conditions
+   * $\vec{n}\times\vec{B}=\vec{n}\times\vec{f}$, then @p first_vector_component would
+   * be 3. The @p boundary_function must return 6 components in this example, with the first 3
+   * corresponding to $\vec{E}$ and the second 3 corresponding to $\vec{B}$.
+   * Vectors are implicitly assumed to have exactly <code>dim</code> components
+   * that are ordered in the same way as we usually order the coordinate directions,
+   * i.e. $x$-, $y$-, and finally $z$-component.
+   * 
+   * The parameter @p boundary_component corresponds to the number @p boundary_indicator
+   * of the face. numbers::internal_face_boundary_id is an illegal value, since it is 
+   * reserved for interior faces.
+   * 
+   * The last argument is denoted to compute the normal vector $\vec{n}$ at the
+   * boundary points.
+   * 
+   *
+   * @ingroup constraints
+   *
+   * @see @ref GlossBoundaryIndicator "Glossary entry on boundary indicators"
+   */
+  template <int dim, class DH>
+  void project_boundary_values_curl_conforming_l2 (const DH &dof_handler,
+                                                   const unsigned int first_vector_component,
+                                                   const Function<dim,double> &boundary_function,
+                                                   const types::boundary_id boundary_component,
+                                                   ConstraintMatrix &constraints,
+                                                   const hp::MappingCollection<dim, dim> &mapping_collection = hp::StaticMappingQ1<dim>::mapping_collection);
```

Comment:
There are only two kinds of DoFHandlers in deal.II -- hp and non-hp. Since you handle the non-hp case below, just remove the second template argument from this function and make it take a hp::DoFHandler as first argument. Then let both this and the other function call an internal function in the .cc file that does what this function currently does.

It's not a big deal but I've found that if something raises a question ("what kinds of DoFHandler can I pass to this function") then one of my students will have this question and it will take 20 minutes to explain :-) The point obviously being that if one of my students has a question, then 10 more around the world will have the same question.